Support bigint during quota evaluation

williamlardier opened the pull request:
Support comparing quotas using bigint, to reduce the impact of handling large numbers.

Issue: CLDSRV-606

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-606 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.41`

* `9.0.0`


Please check the `Fix Version/s` of CLDSRV-606, or the target
branch of this pull request.

benzekrimaha reviewed (approved):
LGTM

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -122,7 +122,7 @@ function _evaluateQuotas(
                     }
                     if (!isMetricStale(bucketMetrics, 'bucket', bucket.getName(), action, inflight, log) &&
-                        bucketMetrics.bytesTotal + inflightForCheck > bucketQuota) {
+                        BigInt(bucketMetrics.bytesTotal || 0) + BigInt(inflightForCheck || 0) > bucketQuota) {
```
should these conversions not be done by `ScubaClient` (when parsing bucketMetrics from result) ?

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -147,8 +147,9 @@ function _evaluateQuotas(
+                    // Metrics are served as BigInt strings
                     if (!isMetricStale(accountMetrics, 'account', account.account, action, inflight, log) &&
-                        accountMetrics.bytesTotal + inflightForCheck > accountQuota) {
+                        BigInt(accountMetrics.bytesTotal || 0) + BigInt(inflightForCheck || 0) > accountQuota) {
```
shoud these conversions not be done by `VaultClient` (when parsing accountMetrics from vault request)?

williamlardier replied:
I need to double check this one: vaultclient uses a unique function to handle response, hence we cannot easily do it. Plus it is also used as a cli tool so we need to ensure this breaks nothing.

francoisferrand commented on lib/api/api.js:
```diff
@@ -272,7 +272,7 @@ const api = {
                 return callback(err);
             }
-            request.accountQuotas = infos?.accountQuota;
+            request.accountQuotas = BigInt(infos?.accountQuota);
```
infos is coming from vault, through vaultclient --> the BigInt conversion should be handled there

francoisferrand commented on tests/utilities/mock/Scuba.js:
```diff
@@ -31,21 +36,21 @@ class Scuba {
+                bytesTotal: ((this._data.bucket.get(bucketName)?.current || 0n) +
+                    (this._data.bucket.get(bucketName)?.nonCurrent || 0n) +
+                    (this._data.bucket.get(bucketName)?.inflight || 0n) +
+                    immediateInflights).toString(),
```
do we need toString() here? Is it not the purpose of `BigInt.prototype.toJSON` above? or should we remove that, if it does not work or is not needed?

Seragonia committed:
Support BigInt notation in S3

Issue: CLDSRV-606

Seragonia committed:
Convert input account quota to bigint for quota evaluation

Issue: CLDSRV-606

Seragonia committed:
Update arsenal version

Issue: CLDSRV-606

Seragonia committed:
Fix java tests

Issue: CLDSRV-606

Seragonia committed:
Support evaluating quotas as bigint

- Value we get from vault is number, but converted into bigint
  internally
- Value from the BucketInfo class is already bigint
- Comparison will use the number received from Scuba as a bigint
- Values may be imprecise but comparing with bigint ensures
  the comparison is correct as per the received values

Issue: CLDSRV-606

williamlardier commented:
/wait

ghost commented:
/approve

benzekrimaha reviewed (approved):
LGTM

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -215,8 +216,9 @@ function validateQuotas(request, bucket, account, apiNames, apiMethod, inflight,
     const requestStartTime = process.hrtime.bigint();
-    const bucketQuota = bucket.getQuota();
-    const accountQuota = account?.quota || 0;
+    const bucketQuota = bucket.getQuota() || 0n;
```
do we have some sort of "functional" test, where the bucketInfo is retrieved from Mongo (and 'parsed' to BigInt) ?
should we wrap this in BigInt() constructor for extra safety? (`BigInt(bucket.getQuota())`) ?

williamlardier replied:
We have this test at the Zenko level, currently. I can update to be safe indeed, but the new BucketInfo class already do it in constructor or `setQuota`?

williamlardier replied:
Actually yes we have the `tests/quota/awsNodeSdk.js` file, it uses a mock for scuba but it really create the buckets and uses the quota APIs to change the quotas, that are then evaluated.

Seragonia committed:
Bump project version

Issue: CLDSRV-606

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-606 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.41`

* `9.0.0`


Please check the `Fix Version/s` of CLDSRV-606, or the target
branch of this pull request.




*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-606` into
`development/8.8`
* [w/9.0/bugfix/CLDSRV-606](https://github.com/scality/cloudserver/compare/development/9.0...w/9.0/bugfix/CLDSRV-606) will be merged into `development/9.0` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.8`

* :heavy_check_mark: `development/9.0`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`



Please check the status of the associated issue CLDSRV-606.

Goodbye williamlardier.




*The following options are set:* **approve**
